opm/simulators/wells/WellInterface_impl.hpp

```diff
@@ -418,6 +418,13 @@ namespace Opm
         // keep a copy of the original well state
         const WellState well_state0 = well_state;
         const double dt = ebosSimulator.timeStepSize();
+        const auto& summary_state = ebosSimulator.vanguard().summaryState();
+        const bool has_thp_limit = this->wellHasTHPConstraints(summary_state);
+        if (has_thp_limit)
+            well_state.well(this->indexOfWell()).production_cmode = Well::ProducerCMode::THP;
+        else
+            well_state.well(this->indexOfWell()).production_cmode = Well::ProducerCMode::BHP;
```

Comment:
It's not directly related to this PR, but why do we save a copy of the `WellState`, modify the original object, and then restore the input object instead of taking the `WellState` by reference-to-`const`, modifying a copy of it as needed, and then letting that copy go out of scope when it's no longer needed?  The latter would save one invocation of `WellState`'s assignment operator.

Reply:
It is a good question. I do not remember how this part was designed originally, with its current way, it looks like the modified/updated `well_state` wants to be used if `converged`. 

But since the control is set forcefully to be `THP` or `BHP` for testing, maybe we should not reuse the updated/modified well state. . 

Then the way you mentioned is the correct way we should use. @bska